### PR TITLE
Implemented FE ACL & updated the examples.

### PR DESCRIPTION
**Pull request by matijaSos**

### CODE REVIEW

File: waspc/data/Generator/templates/react-app/src/router.js [hunk 1]
```diff
@@ -2,6 +2,7 @@
 import React from 'react'
 import { Route, BrowserRouter as Router } from 'react-router-dom'
 
+import createAuthRequiredPage from "./auth/pages/createAuthRequiredPage.js"
```
Comment (Martinsos) [inline waspc/data/Generator/templates/react-app/src/router.js:5]: Why did you import this here?

File: waspc/src/Generator/WebAppGenerator/RouterGenerator.hs [hunk 2]
```diff
@@ -55,9 +66,35 @@ createRouterTemplateData wasp = RouterTemplateData
...
+determineRouteTargetComponent :: Wasp -> Wasp.Route.Route -> String
+determineRouteTargetComponent wasp route =
+    maybe
+        targetPageName 
+        determineRouteTargetComponent'
+        (Wasp.Page._authRequired pageFromRoute)
+    where
+        targetPageName = Wasp.Route._targetPage route
+        -- NOTE(matija): if no page with the name specified in the route, head will fail.
+        pageFromRoute = head $ filter (((==) targetPageName) . Wasp.Page._name) (Wasp.getPages wasp)
```
Comment (Martinsos) [inline waspc/src/Generator/WebAppGenerator/RouterGenerator.hs:87]: I had hard time connecting this with the variable above.
Maybe call them: targetPageName, and targetPage -> then it is obvious. Or, routeTargetPageName and routeTargetPage.

File: waspc/src/Generator/WebAppGenerator/RouterGenerator.hs [hunk 3]
```diff
@@ -55,9 +66,35 @@ createRouterTemplateData wasp = RouterTemplateData
...
+
+        -- | Applied if authRequired property is present.
+        determineRouteTargetComponent' :: Bool -> String
+        determineRouteTargetComponent' authRequired =
+            if authRequired 
+            -- TODO(matija): would be nicer if this function name wasn't hardcoded here.
+            then "createAuthRequiredPage(" ++ targetPageName ++ ")"
```
Comment (Martinsos) [inline waspc/src/Generator/WebAppGenerator/RouterGenerator.hs:94]: How is it that we insert this HOC call directly in the route definition? Why not insert it in the component file? Or do we even have component file, maybe we don't? Maybe this is ok, I am just wondering -> I know before we had special Page files generated, but I am guess we don't create those anymore so it makes sense to do it here, right?

File: waspc/src/Parser/Auth.hs [hunk 4]
```diff
@@ -14,24 +15,52 @@ auth = do
     L.reserved L.reservedNameAuth
     authProperties <- P.waspClosure (L.commaSep1 authProperty)
 
+    let userEntityPs = [s | AuthPropertyUserEntity s <- authProperties]
```
Comment (Martinsos) [inline waspc/src/Parser/Auth.hs:18]: Ps stands for props? Uff it took me some time to figure this out. Why not Props?

File: waspc/src/Parser/Auth.hs [hunk 5]
```diff
@@ -14,24 +15,52 @@ auth = do
...
-        { Wasp.Auth._userEntity = getAuthPropUserEntity authProperties
-        , Wasp.Auth._methods = getAuthPropMethods authProperties
+        { Wasp.Auth._userEntity = head userEntityPs
+        , Wasp.Auth._methods = head methodsPs
+        , Wasp.Auth._onAuthFailedRedirectTo = head redirectPs
         }
 
+-- TODO(matija): this should be extracted if we want to use in other places too.
+failIfPropMissing :: (Applicative m, MonadFail m) => String -> [p] -> m ()
+failIfPropMissing propName ps = when (null ps) $ fail errorMsg
```
Comment (Martinsos) [inline waspc/src/Parser/Auth.hs:35]: Aha fail, nice that works well in context of Parser, returns line?

**Comment by Martinsos** [comment]
@matijaSos Tests are not passing!

File: examples/realworld/ext/article/components/ArticleEditorPage.js [hunk 6]
```diff
@@ -12,26 +11,18 @@ import getArticle from '@wasp/queries/getArticle'
...
-    return <span> Please <Link to='/login'>log in</Link>. </span>
-  }
-
   return articleError
     ? articleError.message || articleError
     : (
       <div>
         <Navbar />
-        <ArticleEditor user={user} article={article} />
+        <ArticleEditor user={props.user} article={article} />
```
[the remark this answers is not in the document]
Comment (matijaSos) [reply examples/realworld/ext/article/components/ArticleEditorPage.js:25]: cool, added prop types.

File: waspc/data/Generator/templates/react-app/src/auth/pages/createAuthRequiredPage.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    if (isSuccess) {
+      if (user) {
+        return (
+          <Page {...props} user={user} />
+        )
+      } else {
+        return <Redirect to="{= onAuthFailedRedirectTo =}" />
+      }
+    } else if (isLoading) {
+      return <span>Loading...</span>
```
[the remark this answers is not in the document]
Comment (matijaSos) [reply waspc/data/Generator/templates/react-app/src/auth/pages/createAuthRequiredPage.js:21]: yes, makes sense!

File: waspc/src/Generator/WebAppGenerator/RouterGenerator.hs (hunk 3 above)
Comment (matijaSos) [reply waspc/src/Generator/WebAppGenerator/RouterGenerator.hs:94]: yes, here I am taking directly a page component provided by user and I wrap it in createAuthRequiredPage. We could also generate new files in which HOC takes their page in, haven't thought of it before. Both ways sound good to me, but this one (one currently implemented) is simpler?

File: waspc/src/Parser/Auth.hs (hunk 4 above)
Comment (matijaSos) [reply waspc/src/Parser/Auth.hs:18]: we are in Haskell come one :D

File: waspc/src/Generator/WebAppGenerator/RouterGenerator.hs (hunk 2 above)
Comment (matijaSos) [reply waspc/src/Generator/WebAppGenerator/RouterGenerator.hs:87]: makes sense, changed it to targetPage.

File: waspc/src/Parser/Auth.hs (hunk 5 above)
Comment (matijaSos) [reply waspc/src/Parser/Auth.hs:35]: yes! although it is not the exact line, it is the one of the next closure (I guess it is connected to what amount of input has already been consumed by parser), but still better than nothing.

File: waspc/src/Parser/Auth.hs (hunk 4 above)
Comment (matijaSos) [reply waspc/src/Parser/Auth.hs:18]: changed to Props

**Pull request merged**